src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py

```diff
@@ -336,7 +336,14 @@ def _encode_prompt(
             )
             prompt_embeds = prompt_embeds[0]
 
-        prompt_embeds = prompt_embeds.to(dtype=self.text_encoder.dtype, device=device)
+        if self.text_encoder is not None:
+            prompt_embeds_dtype = self.text_encoder.dtype
+        elif self.unet is not None:
+            prompt_embeds_dtype = self.unet.dtype
+        else:
+            prompt_embeds_dtype = prompt_embeds.dtype
+
+        prompt_embeds = prompt_embeds.to(dtype=prompt_embeds_dtype, device=device)
```

Comment:
I get that when this method is responsible for encoding the prompt, it gets it from `text_encoder` -- so converting things to `text_encoder`'s data type is a way to enforce consistency between the different modes.

On the other hand, the output is being used as conditioning data for the unet. In which case the type of the encoding network doesn't really matter, right? This data is never going back in to that network, only onward to the unet. Is there any reason to not simplify that conditional somewhat and drop the reference to `text_encoder.dtype` entirely?

Reply:
Looking in the git history, https://github.com/huggingface/diffusers/commit/b562b6611fb53dae9bcffcaaf44d944539ae22de this was the commit it was introduced. I think the intention for the cast was that when prompt_embeds were passed, they would need to be cast to the equivalent type as if they had been produced by the text encoder.

I think ideally the cleanest solution here would be to just cast to the type of the unet, but we also have the case of if the pipeline doesn't have a unet, i.e. if the user is just instantiating the pipeline to encode text (not often but it's documented in some of the example IF pipeline workflows). So I think while not ideal this might be the simplest way to cover all bases